### DISCUSSION TITLE
Inserting quiz in quarto document

### DISCUSSION THREAD

**Discussion by rputikar** [quarto-dev/quarto-cli#5115] [Q&A]
Hello,

I am interested in inserting a quiz (rendered using something like surveyjs) in a quarto document. Any pointers on how to get it working would be highly appreciated. 

thanks, Ranjeet.

**Comment by mcanouil** [comment]
I would suggest you embed the JavaScript library as a Quarto extension, see <https://quarto.org/docs/extensions/creating.html>.

**Comment by baptiste** [comment]
I recently noticed that this [online book](https://jverzani.github.io/CalculusWithJuliaNotes.jl/derivatives/lhospitals_rule.html) features a little quiz at the end of many pages; it uses [QuizQuestions.jl](https://github.com/jverzani/QuizQuestions.jl) so it's relying on Julia, but that might be an inspiration even if you're using a different language.

**Comment by HeidiSeibold** [comment]
I was just researching the same question and came accross this blogpost: https://web.mat.upc.edu/joaquim.puig/posts/webexercises-quiz/

It uses the R package [webexercises](https://psyteachr.github.io/webexercises/).

**Comment by joelostblom** [comment]
Someone made a quarto extension for this https://github.com/nareal/naquiz. I'm curious to hear if there are any plans to support something like this natively in Quarto?

**Reply by mcanouil** [reply]
Cc @andrewpbray

**Reply by cscheid** [reply]
No plans.

**Comment by saskiaotto** [comment]
For anyone still looking for a solution: I recently published the R package
**rquiz** on CRAN, which provides interactive quizzes as HTML widgets that
work natively in R Markdown and Quarto documents (HTML, Slidy, RevealJS).
 
It offers three quiz types:
 
- **Single-question quiz** with instant visual feedback
- **Multi-page quiz** with navigation, timer, and results summary
- **Fill-in-the-blank (cloze)** exercises
A minimal example in a `.qmd` file:
 
```r
library(rquiz)
 
singleQuestion(
  question = "Which function reads a CSV file in base R?",
  options = c("read.csv()", "load.csv()", "import.csv()", "open.csv()"),
  answer = 1
)
```
 
No Shiny server needed — it's pure Vanilla JavaScript, so it works in any
static HTML output. It also supports customizable theming, multilingual UI
(EN/DE/FR/ES), answer shuffling, and keyboard/ARIA accessibility.
 
- CRAN: https://cran.r-project.org/package=rquiz
- Documentation & live demos: https://saskiaotto.github.io/rquiz/
- Quarto demo files (HTML, Slidy, RevealJS): https://saskiaotto.github.io/rquiz/articles/slides-demo.html
I've been using these quizzes in my university lectures since 2020 and
finally packaged them up. Happy to answer questions!

**Reply by mcanouil** [reply]
Thanks for sharing!

**Comment by mcanouil** [comment]
For reference:

- https://github.com/parmsam/quarto-quizdown
- https://github.com/parmsam/quarto-quiz
- https://surveydown.org/

**Reply by saskiaotto** [reply]
Thanks for the references! Those are great tools, I wasn't aware of them.
rquiz comes at it from a different angle: it's an R package that creates self-contained htmlwidgets, so quizzes work anywhere htmlwidgets work (R Markdown, Quarto HTML/Slidy/RevealJS, Shiny, RStudio Viewer) without needing a server, a database, or a Quarto extension install. The focus is on quick, visual knowledge checks in teaching materials rather than data-collecting surveys. But I'll definitely check them out — always good to know what's available in this space!
